cuda/base/math.hpp

```diff
@@ -93,17 +94,126 @@ __device__ GKO_INLINE std::complex<double> one<std::complex<double>>()
+    // integer representation of a floating point number, where all exponent
+    // bits are set
+    static constexpr int_type exponent_mask =
+        ((int_type{1} << number_exponent_bits) - 1) << number_significand_bits;
+    static __device__ int_type reinterpret_int(const double &value)
+    {
+        return __double_as_longlong(value);
+    }
+};
+
 
-// For some CUDA versions, the complex variation for `isfinite` has to be
-// manually defined
-#if defined(GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE)
+}  // namespace detail
 
-GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE(thrust::complex<double>);
-GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE(thrust::complex<float>);
 
-#endif  // defined(GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE)
+/**
+ * Checks if a given value is finite, meaning it is neither +/- infinity
+ * nor NaN.
+ *
+ * @internal  It checks if all exponent bits are set. If all are set, the
+ *            number either represents NaN or +/- infinity, meaning it is a
+ *            non-finite number.
+ *
+ * @param value  value to check
+ *
+ * returns `true` if the given value is finite, meaning it is neither
+ *         +/- infinity nor NaN.
+ */
+#define GKO_DEFINE_ISFINITE_FOR_TYPE(_type)                               \
+    GKO_INLINE __device__ bool isfinite(const _type &value)               \
+    {                                                                     \
+        constexpr auto mask = detail::mask_creator<_type>::exponent_mask; \
+        const auto re_int =                                               \
+            detail::mask_creator<_type>::reinterpret_int(value);          \
+        return (re_int & mask) != mask;                                   \
+    }
+
+GKO_DEFINE_ISFINITE_FOR_TYPE(float)
+GKO_DEFINE_ISFINITE_FOR_TYPE(double)
+#undef GKO_DEFINE_ISFINITE_FOR_TYPE
+
+
+/**
+ * Checks if all components of a complex value are finite, meaning they are
+ * neither +/- infinity nor NaN.
+ *
+ * @internal required for the clang compiler. This function will be used rather
+ *           than the `isfinite` function in the public `math.hpp` because
+ *           there is no template parameter, so it is prefered during lookup.
+ *
+ * @tparam T  complex type of the value to check
+ *
+ * @param value  complex value to check
+ *
+ * returns `true` if both components of the given value are finite, meaning
+ *         they are neither +/- infinity nor NaN.
+ */
+#define GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE(_type)              \
+    GKO_INLINE __device__ bool isfinite(const _type &value)      \
+    {                                                            \
+        return isfinite(value.real()) && isfinite(value.imag()); \
+    }
+
+GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE(thrust::complex<float>)
+GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE(thrust::complex<double>)
+#undef GKO_DEFINE_ISFINITE_FOR_COMPLEX_TYPE
+
+
+// For all other compiler in combination with CUDA, just use the provided
+// `isfinite` function
+#elif defined(__CUDA_ARCH__)
```

Comment:
Since you're in the CUDA module a `#else` should be enough now?

Reply:
Wait, scrap that comment. What you want here is to have this only for device code, which is why you have this.

Reply:
No, because some parts are still host code, which would not have the `__CUDA_ARCH__` defined, and which should use `std::isfinite` (`::isfinite` does not exist for host code, resulting in a compiler error on some systems).

Reply:
Exactly